### PR TITLE
replace computeNestedModel and computeAttributeReference with computeAttribute

### PR DESCRIPTION
**Pull request by igorT**

### CODE REVIEW

File: addon/record-data.js [hunk 1]
```diff
@@ -42,6 +43,16 @@ class M3SchemaInterface {
     this._suppressNotifications = false;
   }
 
+  nested(object) {
+    schemaTypesInfo.set(object, NESTED);
+    return object;
+  }
```
Comment (hjdivad) [inline addon/record-data.js:49]: comment this public api

File: addon/record-data.js [hunk 2]
```diff
@@ -42,6 +43,16 @@ class M3SchemaInterface {
...
+
+  reference(object) {
+    schemaTypesInfo.set(object, REFERENCE);
+    return object;
+  }
```
Comment (hjdivad) [inline addon/record-data.js:54]: comment this public api

File: addon/record-data.js (hunk 1 above)
Comment (hjdivad) [reply addon/record-data.js:49]: also add docs in readme

File: addon/resolve-attribute-util.js [hunk 3]
```diff
@@ -114,13 +116,32 @@ export function resolveValue(key, value, modelName, store, schema, record, paren
   const schemaInterface = recordData.schemaInterface;
 
   // First check to see if given value is either a reference or an array of references
```
Comment (hjdivad) [inline addon/resolve-attribute-util.js:118]: this comment is now out of place

File: addon/resolve-attribute-util.js [hunk 4]
```diff
@@ -114,13 +116,32 @@ export function resolveValue(key, value, modelName, store, schema, record, paren
...
-  let reference = computeAttributeReference(key, value, modelName, schemaInterface, schema);
-  if (reference !== undefined && reference !== null) {
+  let computedValue, nested, reference;
+  if (schema.useComputeAttribute()) {
```
Comment (hjdivad) [inline addon/resolve-attribute-util.js:120]: this is a hot path; at a minimum we will want to cache this

File: addon/resolve-attribute-util.js [hunk 5]
```diff
@@ -114,13 +116,32 @@ export function resolveValue(key, value, modelName, store, schema, record, paren
...
+    computedValue = computeAttribute(key, value, modelName, schemaInterface, schema);
+    if (computedValue && typeof computedValue === 'object') {
```
Comment (hjdivad) [inline addon/resolve-attribute-util.js:122]: should confirm that this is the best way to check for POJOness in a hot path

File: addon/services/m3-schema-manager.js [hunk 6]
```diff
@@ -21,7 +23,16 @@ export default class SchemaManager extends Service {
...
+    let schema = this.get('schema');
+    if (this.useComputeAttribute()) {
+      let computedValue = schema.computeAttribute(key, value, modelName, schemaInterface);
+      let valueType = schemaTypesInfo.get(computedValue);
+      if (valueType === REFERENCE) {
+        return computedValue;
+      }
+    } else {
+      return schema.computeAttributeReference(key, value, modelName, schemaInterface);
+    }
```
Comment (hjdivad) [inline addon/services/m3-schema-manager.js:35]: I would not expect someone opted in to the new API to end up with call paths going through the old API.  Is there a good reason for this?

File: addon/services/m3-schema-manager.js [hunk 7]
```diff
@@ -35,7 +46,30 @@ export default class SchemaManager extends Service {
...
+    if (this.useComputeAttribute()) {
+      let computedValue = schema.computeAttribute(key, value, modelName, schemaInterface);
+      let valueType = schemaTypesInfo.get(computedValue);
+      if (valueType === NESTED) {
+        return computedValue;
+      }
+    } else {
+      return schema.computeNestedModel(key, value, modelName, schemaInterface);
+    }
+  }
```
Comment (hjdivad) [inline addon/services/m3-schema-manager.js:59]: I would not expect someone opted in to the new API to end up with call paths going through the old API.  Is there a good reason for this?

File: addon/services/m3-schema-manager.js [hunk 8]
```diff
@@ -35,7 +46,30 @@ export default class SchemaManager extends Service {
...
+
+  useComputeAttribute() {
+    let schema = this.get('schema');
+    let isComputeAttributeReferenceMissing = !schema.computeAttributeReference;
+    let defaultPrototype = DefaultSchema.prototype;
+    let usingDefaultSchemaMethods =
+      defaultPrototype.computeAttributeReference === schema.computeAttributeReference &&
+      defaultPrototype.computeNestedModel === schema.computeNestedModel;
+    return isComputeAttributeReferenceMissing || usingDefaultSchemaMethods;
```
Comment (hjdivad) [inline addon/services/m3-schema-manager.js:72]: At a minimum we'll need to cache this

Comment (hjdivad) [inline addon/services/m3-schema-manager.js:72]: It makes more sense to me to check against `computeAttribute` being implemented than the inverse

File: package.json [hunk 9]
```diff
@@ -65,7 +65,7 @@
   "dependencies": {
     "babel-plugin-debug-macros": "^0.3.3",
     "broccoli-funnel": "^3.0.3",
-    "ember-cli-babel": "^7.21.0",
+    "ember-cli-babel": "7.18",
```
Comment (hjdivad) [inline package.json:68]: This seems suspect.  We should probably remove it.

File: addon/resolve-attribute-util.js (hunk 4 above)
Comment (igorT) [reply addon/resolve-attribute-util.js:124]: :+1: cached in a weak map

File: addon/resolve-attribute-util.js (hunk 5 above)
Comment (igorT) [reply addon/resolve-attribute-util.js]: good catch, I asked and it turned out that checking for null explicitly is better, fixed

File: addon/services/m3-schema-manager.js (hunk 6 above)
Comment (igorT) [reply addon/services/m3-schema-manager.js]: Good point, fixed

File: addon/services/m3-schema-manager.js (hunk 8 above)
Comment (igorT) [reply addon/services/m3-schema-manager.js]: Cached using a weakmap

Comment (igorT) [reply addon/services/m3-schema-manager.js]: Fixed

File: package.json (hunk 9 above)
Comment (igorT) [reply package.json]: Yup, it was leftover from testing, as linking was unhappy with babel issues at some point

File: tests/unit/model/changed-attrs-test.js [hunk 10]
```diff
@@ -1284,6 +1319,7 @@ module('unit/model/changed-attrs', function(hooks) {
     );
   });
 
+  /*
```
Comment (igorT) [inline tests/unit/model/changed-attrs-test.js:1322]: Test not valid for the new world, will bring back in the old tests

File: addon/resolve-attribute-util.js (hunk 3 above)
Comment (igorT) [reply addon/resolve-attribute-util.js]: fixed

File: addon/services/m3-schema-manager.js (hunk 7 above)
Comment (igorT) [reply addon/services/m3-schema-manager.js]: Good point, fixed

File: addon/record-data.js (hunk 2 above)
Comment (igorT) [reply addon/record-data.js:63]: commented

File: addon/record-data.js (hunk 1 above)
Comment (igorT) [reply addon/record-data.js:54]: done

File: tests/unit/model/state-test.js [hunk 11]
```diff
@@ -6,138 +6,193 @@ import { isArray } from '@ember/array';
...
-        } else if (Array.isArray(refValue)) {
-          return refValue.map(x => ({
-            type: null,
-            id: x,
-          }));
-        }
-        return null;
-      }
+let computeNestedModel = function computeNestedModel(key, value) {
+  if (value && typeof value === 'object' && value.constructor !== Date && !isArray(value)) {
```
Comment (runspired) [inline tests/unit/model/state-test.js:10]: I believe we can clean up this check by removing the Date check (seems no date usage below).

File: tests/unit/model/tracked-array-test.js [hunk 12]
```diff
@@ -7,297 +7,336 @@ import DefaultSchema from 'ember-m3/services/m3-schema';
...
-        }
-
-        computeNestedModel(key, value /*, modelName, schemaInterface */) {
-          if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
-            return {
-              attributes: value,
-            };
-          }
+function computeNestedModel(key, value /*, modelName, schemaInterface */) {
+  if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
```
Comment (runspired) [inline tests/unit/model/tracked-array-test.js:11]: we do effectively this check a lot of places in both our code and our tests, and it seems slightly different every time. This is likely to be a hot path in our actual code, so we should audit at some point to ensure that the simplest bailout is being done prior to the typeof check.

File: tests/unit/projection-test.js [hunk 13]
```diff
@@ -1296,1828 +1404,1830 @@ module('unit/projection', function(hooks) {
...
-          data: {
-            id: BOOK_ID,
-            type: BOOK_EXCERPT_PROJECTION_CLASS_PATH,
-            attributes: {
-              title: BOOK_TITLE_2,
-            },
-          },
-        });
-      });
+  module('creating/updating projections with resolutions', function(hooks) {
```
Comment (runspired) [inline tests/unit/projection-test.js:3057]: lets split this module out and make sure we test the old hooks (removed accidentally)

File: tests/unit/model/state-test.js (hunk 11 above)
Comment (igorT) [reply tests/unit/model/state-test.js:10]: Cleaned up

File: tests/unit/projection-test.js (hunk 13 above)
Comment (igorT) [reply tests/unit/projection-test.js]: done

**Comment by igorT** [comment]
@hjdivad addressed all of the feedback, and @runspired gave it a read through, so will be merging

**Pull request merged**